### PR TITLE
Fix: [Clear Button is active when no input is selected in Dropdown]

### LINKED ISSUE #9263: Clear button is active when no input is selected in dropdown
**Describe the bug**

AutoComplete dropdown clear-button state remains show always even when there are no selected field.  When clicked, it opens the dropdown and when the user clicks outside it then clears out.

**To Reproduce**

1. Go to Patients
2. Click on add a patient
3. When the prompt is open, the clear button is already present even when the user hasn't selected anything.

**Expected behavior**

- The clear button state shouldn't been shown when no input field is selected.

**Screenshots**

https://github.com/user-attachments/assets/a606cfde-a606-4d6e-9b72-290a3a325d61

**Desktop (please complete the following information):**

- OS: MacOS
- Browser - Chrome

### PR DESCRIPTION
**Pull request by Rustix69**
## Proposed Changes

- Fixes #9263 
- Added a state const [hasTyped, setHasTyped] = useState(false);.
- Modified handleInputChange to set hasTyped(true) when the user types
- In handleSelectionChange, set setQuery(optionLabel(selectedOption)); to update query after selection.
- Set hasTyped(true) after selection to ensure the "Clear Selection" button is shown.
- In the ComboboxButton rendering logic, updated the condition for showing the "Clear Selection" button:
- Added setHasTyped(false) in the onClick handler for clearing selection to reset the hasTyped state.

## Behaviour after the changes
- The website allows users to search and select options from a dropdown, with the input dynamically filtering results as the user types.
- The "Clear Selection" button only appears after the user types or selects an option, making it easy to clear the selected value.
- The input field reflects the current query, and the dropdown updates based on the user's input, offering a smooth and interactive search experience.

### Screen Records
[Screencast from 2024-12-02 00-14-08.webm](https://github.com/user-attachments/assets/75ec8458-6b82-4c94-9c2a-099f4bd1f832)

## Summary by CodeRabbit

- **New Features**
	- Enhanced facility selection handling allows for a clearer representation of no facility selected.
	- Improved navigation logic ensures actions only proceed with a valid facility selected.

- **Bug Fixes**
	- Error handling added to prevent navigation attempts without a selected facility.

- **Documentation**
	- Updated type definitions for state management to reflect nullable values for better clarity.

### CODE REVIEW

**Comment by Rustix69** [comment]
@Jacobjeevan I had not touched the other codes but its giving errors in Cypress Tests that are being shown in different components. And the clear button is working smoothly after making the code changes

File: src/components/Form/AutoCompleteAsync.tsx [hunk 1]
```diff
@@ -128,14 +144,16 @@ const AutoCompleteAsync = (props: Props) => {
             {!disabled && (
               <ComboboxButton className="absolute inset-y-0 right-0 flex items-center pr-2">
                 <div className="absolute right-0 mr-2 flex items-center text-lg text-secondary-900">
-                  {hasSelection && !loading && !required && (
+                  {shouldShowClearButton && !loading && !required && (
```
Comment (Jacobjeevan) [inline src/components/Form/AutoCompleteAsync.tsx]: Let's not change AutoCompleteAsync.

You will notice that issue doesn't persist across all FacilitySelect instances, but rather in ManagePatients and AssetsList; hasSelection normally takes care of this, so look into those two instances and fix those.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

**Comment by Jacobjeevan** [comment]
@ANIR1604 Any updates?

**Comment by Rustix69** [comment]
> @ANIR1604 Any updates?

Now It's working Perfect Sir @Jacobjeevan. I had attached a screencast of the website after the changes. Plz review and approve my PR if it's ok.


[Screencast from 2024-12-06 03-55-34.webm](https://github.com/user-attachments/assets/24d99976-d9c9-4426-86bc-bd7901ef14d1)

File: src/components/Assets/AssetsList.tsx [hunk 2]
```diff
@@ -478,12 +477,12 @@ const AssetsList = () => {
       )}
       {facility && (
         <AssetImportModal
-          open={importAssetModalOpen}
+          open={importAssetModalOpen} // Reset selected facility to null on cancel
```
Comment (Jacobjeevan) [inline src/components/Assets/AssetsList.tsx]: Remove the comments

File: src/components/Patient/ManagePatients.tsx [hunk 3]
```diff
@@ -958,18 +957,25 @@ export const PatientManager = () => {
...
+          if (selectedFacility) {
+            switch (showDialog) {
+              case "create":
+                navigate(`facility/${selectedFacility.id}/patient`);
+                break;
+              case "list-discharged":
+                navigate(`facility/${selectedFacility.id}/discharged-patients`);
+                break;
+              default:
+                console.error("Unknown dialog state");
```
Comment (Jacobjeevan) [inline src/components/Patient/ManagePatients.tsx]: This isn't needed here.

File: src/components/Patient/ManagePatients.tsx [hunk 4]
```diff
@@ -958,18 +957,25 @@ export const PatientManager = () => {
...
+                break;
+            }
+          } else {
+            console.error("No facility selected");
```
Comment (Jacobjeevan) [inline src/components/Patient/ManagePatients.tsx]: Replace with the Notifications instead.

File: src/components/Assets/AssetsList.tsx [hunk 5]
```diff
@@ -496,10 +495,18 @@ const AssetsList = () => {
...
         setSelected={(e) => setSelectedFacility(e)}
         selectedFacility={selectedFacility}
-        handleOk={() => navigate(`facility/${selectedFacility.id}/assets/new`)}
+        handleOk={() => {
+          if (selectedFacility) {
+            // Proceed with navigation if selectedFacility is not null
+            navigate(`facility/${selectedFacility.id}/assets/new`);
+          } else {
+            // Handle case when selectedFacility is null (e.g., show an error message)
+            console.error("No facility selected");
```
Comment (Jacobjeevan) [inline src/components/Assets/AssetsList.tsx]: Replace with Notifications instead.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Assets/AssetsList.tsx (hunk 2 above)
Comment (Jacobjeevan) [reply src/components/Assets/AssetsList.tsx]: Also check the behavior of clear selection on the popup for Import/Export. It's not working as intended.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

**Comment by Rustix69** [comment]
@Jacobjeevan Solved all the errors related to clear selection at all instances.

File: src/components/Patient/ManagePatients.tsx [hunk 6]
```diff
@@ -966,11 +966,11 @@ export const PatientManager = () => {
                 navigate(`facility/${selectedFacility.id}/discharged-patients`);
                 break;
               default:
-                console.error("Unknown dialog state");
+                Notification.Error({ msg: "Unknown dialog state" });
```
Comment (Jacobjeevan) [inline src/components/Patient/ManagePatients.tsx]: You missed one of the earlier comments (make sure to resolve each one, helps with keeping track as well).

No need for a default state here 🤔 

Also you don't need to request review and tag - one of those is enough 😅

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Patient/ManagePatients.tsx (hunk 6 above)
Comment (Rustix69) [reply src/components/Patient/ManagePatients.tsx]: Ok Sir, Can we connect in LinkedIn ? It will be a wonderful learning curve for me to work with you sir

**Review by Jacobjeevan — APPROVED** [review]

**Comment by nihal467** [comment]
LGTM

File: src/components/Patient/ManagePatients.tsx [hunk 7]
```diff
@@ -100,9 +100,8 @@ export const PatientManager = () => {
       "emergency_phone_number",
     ],
   });
-  const [selectedFacility, setSelectedFacility] = useState<FacilityModel>({
-    name: "",
-  });
+  const [selectedFacility, setSelectedFacility] =
+    useState<FacilityModel | null>(null);
```
Comment (rithviknishad) [inline src/components/Patient/ManagePatients.tsx]:
```suggestion
  const [selectedFacility, setSelectedFacility] = useState<FacilityModel>();
```

File: src/components/Assets/AssetsList.tsx [hunk 8]
```diff
@@ -46,15 +46,14 @@ const AssetsList = () => {
   const [isLoading, setIsLoading] = useState(false);
   const [isScannerActive, setIsScannerActive] = useState(false);
   const [totalCount, setTotalCount] = useState(0);
-  const [facility, setFacility] = useState<FacilityModel>();
+  const [facility, setFacility] = useState<FacilityModel | null>(null);
```
Comment (rithviknishad) [inline src/components/Assets/AssetsList.tsx]:
```suggestion
  const [facility, setFacility] = useState<FacilityModel>();
```

File: src/components/Assets/AssetsList.tsx [hunk 9]
```diff
@@ -46,15 +46,14 @@ const AssetsList = () => {
...
   const [status, setStatus] = useState<string>();
   const [asset_class, setAssetClass] = useState<string>();
   const [importAssetModalOpen, setImportAssetModalOpen] = useState(false);
   const assetsExist = assets.length > 0 && Object.keys(assets[0]).length > 0;
   const [showFacilityDialog, setShowFacilityDialog] = useState(false);
-  const [selectedFacility, setSelectedFacility] = useState<FacilityModel>({
-    name: "",
-  });
+  const [selectedFacility, setSelectedFacility] =
+    useState<FacilityModel | null>(null);
```
Comment (rithviknishad) [inline src/components/Assets/AssetsList.tsx]:
```suggestion
  const [selectedFacility, setSelectedFacility] = useState<FacilityModel>();
```

File: src/components/Assets/AssetsList.tsx [hunk 10]
```diff
@@ -460,15 +459,11 @@ const AssetsList = () => {
           </div>
         </>
       )}
-      {typeof facility === "undefined" && (
+      {facility === null && (
```
Comment (rithviknishad) [inline src/components/Assets/AssetsList.tsx]: `==` on null will check against null and undefined

```suggestion
      {facility == null && (
```

**Comment by Rustix69** [comment]
@rithviknishad Ok I will be working on this tomorrow. Can't I change it to null at Time 0 ? In my point of view it will be a great approach to give it a null value.

**Comment by Rustix69** [comment]
@rithviknishad Sir plz review my PR ... Are we ok to go with these changes ?

**Review by rithviknishad — APPROVED** [review]

**Pull request merged**